### ISSUE TITLE
cross validation in machine learning

### ISSUE THREAD

**Issue by reedwave** [reedwave/note#4]
## 交叉验证的3个集合

* 划分多种集合，目的即是为了提高模型的泛化能力，既不欠拟合也不过拟合，通常有3种集合
* 训练集(`train set`)：用于训练模型的数据集
* 验证集(`validation set`)：用于模型调参的数据集，优化模型
* 测试集(`test set`)：用于检验已确定模型的能力
* 上述3种必然涉及到训练集和测试集，很多情况下并涉及不到验证集。

## `train_test_split(*array, **options)`

`sklearn.model_selection.train_test_split`划分出训练集和测试集

参数:

    *array: 需要拆分的数据集
    test_size: 拆分成测试集的数据比例，0-1之间的数字，默认值为0.25
    train_size: 拆分成训练集的数据比例，默认值为1-test_size
    random_state: 随机数种子，如果不指定，则每次随机出的训练集和测试集都是不同，如果指定则每次拆分出的训练集和测试集都一样

#### 例子:

这样我们能快速的取出70%作为训练集，和30%的测试集

```py
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_iris
iris = load_iris()
train, target = iris.data, iris.target
x_train, x_test, y_train, y_test = train_test_split(train, target, test_size=0.3, random_state=0)
train.shape, x_train.shape, x_test.shape

>>> ((150, 4), (105, 4), (45, 4))
```

## `KFold(n_splits=3, shuffle=False, random_state=None)`
`sklearn.model_selection.KFold`K折交叉验证，将数据集划分为k个较小的子集，将k-1个子集作为训练集，将剩余的1份作为测试集，k折交叉验证得出的性能指标是循环计算中每个值的平均值，在处理样本数据集较少的问题时比较有优势

参数:

    n_splits    : 折数，最少为2
    shuffle     : 是否随机 True or False
    random_state: 同上随机种子，当shuffle==True时会使用此参数

例子:

KFold拆分出来的是数据集的索引

```py
from sklearn.model_selection import KFold
import numpy as np
kf = KFold()
train = np.array([[1, 2], [3, 4], [5, 6], [7, 8], [9, 10], [11, 12]])
for train_index, test_index in kf.split(train):
    print train_index, test_index

>>> [2 3 4 5] [0 1]
    [0 1 4 5] [2 3]
    [0 1 2 3] [4 5]
```

## 计算交叉验证的指标
`cross_val_score(estimator, X, y=None, groups=None, scoring=None, cv=None, n_jobs=1, verbose=0, fit_params=None, pre_dispatch='2*n_jobs')`

经过`KFold`方法拆分的数据集可以用`cross_val_score`初次简单的计算模型的分数

参数:

    estimator: fit之后的预测器
    X        : 测试的数据
    Y        : 测试数据的正确答案
    scoring  : 计算结果分数的类型
    cv       : KFold折数
    n_jobs   : CPU使用数，-1代表所有CPU

```py
from sklearn.model_selection import cross_val_score
from sklearn.svm import SVC
clf = SVC(kernel='linear', C=1)
scores = cross_val_score(clf, iris.data, iris.target, cv=5, scoring='precision_macro')
print(scores)

>>> [ 0.96969697  1.          0.96969697  0.96969697  1.        ]
```
